src/encodings/encoding.rs

```diff
@@ -630,6 +636,136 @@ impl DeltaBitPackEncoderConversion<Int64Type> for DeltaBitPackEncoder<Int64Type>
   }
 }
 
+/// Encoding for byte arrays to separate the length values and the data.
+/// The lengths are encoded using DELTA_BINARY_PACKED encoding, data is
+/// stored as raw bytes.
+pub struct DeltaLengthByteArrayEncoder<T: DataType> {
+  // length encoder
+  len_encoder: DeltaBitPackEncoder<Int32Type>,
+  // byte array data
+  data: Vec<ByteArray>,
+  _phantom: PhantomData<T>
+}
+
+impl<T: DataType> DeltaLengthByteArrayEncoder<T> {
+  pub fn new() -> Self {
+    Self {
+      len_encoder: DeltaBitPackEncoder::new(),
+      data: vec!(),
+      _phantom: PhantomData
+    }
+  }
+}
+
+default impl<T: DataType> Encoder<T> for DeltaLengthByteArrayEncoder<T> {
+  fn put(&mut self, _values: &[T::T]) -> Result<()> {
+    panic!("DeltaLengthByteArrayEncoder only supports ByteArrayType");
+  }
+
+  fn encoding(&self) -> Encoding {
+    Encoding::DELTA_LENGTH_BYTE_ARRAY
+  }
+
+  fn flush_buffer(&mut self) -> Result<ByteBufferPtr> {
+    panic!("DeltaLengthByteArrayEncoder only supports ByteArrayType");
+  }
+}
+
+impl Encoder<ByteArrayType> for DeltaLengthByteArrayEncoder<ByteArrayType> {
+  fn put(&mut self, values: &[ByteArray]) -> Result<()> {
+    let lengths: Vec<i32> =
+      values.iter().map(|byte_array| byte_array.len() as i32).collect();
+    self.len_encoder.put(&lengths)?;
+    for byte_array in values {
+      self.data.push(byte_array.clone());
```

Comment:
I assume that clone for byte array, triggers clone for byte buffer ptr, which only clones Rc, not the actual data.

Reply:
@sadikovi yes that's right. It will not clone the actual data.